junit/src/main/java/cucumber/runtime/junit/IdProvider.java

```diff
@@ -0,0 +1,56 @@
+package cucumber.runtime.junit;
+
+import java.io.Serializable;
+
+/**
+ * JUnit Descriptions require an unique id. However this id should be identical between test
+ * re-runs. This allows descriptions to be used to determine which failed tests should be rerun.
+ *
+ * IdProvider generates predictable unique ids. The feature name is used as a base to disambiguate
+ * between different features.
+ */
+class IdProvider {
+
+    private long id = 0;
+    private final String base;
+
+    IdProvider(String base) {
+        this.base = base;
+    }
+
+    Serializable next() {
+        return new Id(base, id++);
```

Comment:
We're numbering the tree of Description elements in depth first order. Because the description elements are only created once and cached afterwards this is sufficient to identify which scenario needs to be rerun.

For example:

```feature
Feature: Basic Arithmetic
  Background: A Calculator
    Given a calculator I just turned on

  Scenario: Addition
    When I add 4 and 5
    Then the result is 9

  Scenario: Another Addition
    When I add 4 and 7
    Then the result is 11
```

Results in a tree of executors that looks like this:

```
FeatureRunner => Feature: Basic Arithmetic
  |- ScenarioRunner => Scenario: Addition
  |   |- ExecutionUnitRunner => 
  |       |- Background step: Given a calculator I just turned on
  |       |- When I add 4 and 5
  |       |- Then the result is 9
  |- ScenarioRunner => Scenario: Another Addition
    |- ExecutionUnitRunner => 
        |- Background step: Given a calculator I just turned on
        |- When I add 4 and 7
        |- Then the result is 11
```
This tree can then be numbered in depth first order:
```
0. FeatureRunner => Feature: Basic Arithmetic
1.   |- ScenarioRunner => Scenario: Addition
2.  |   |- ExecutionUnitRunner => 
2.  |       |- Background step: Given a calculator I just turned on
4.  |       |- When I add 4 and 5
5.  |       |- Then the result is 9
6.  |- ScenarioRunner => Scenario: Another Addition
7.    |- ExecutionUnitRunner => 
8.        |- Background step: Given a calculator I just turned on
9.        |- When I add 4 and 7
10.       |- Then the result is 11
```

Which means that for a given feature file creating the numbering of descriptions is entirely predictable. So I guess I don't quite understand your comment.

> The id should instead be base on the uri(/path) of the feature file and the line number of the feature/scenario/etc.

Using the feature uri and the line number alone is not feasible as background steps are repeated.  A failing background step in the execution of Scenario `Another Addition` would result in both `Another Addition` and `Addition` being rerun. Their descriptions both contain a child Description that uses line 3 .

It would be possible to use the uri and line number. But then we'd need some other other identifier to disambiguate between background steps in different scenarios. So as far as I can see numbering the tree by depth first order is sufficient and simple.